### DISCUSSION TITLE
[2023/06/06] Frontend Night

### DISCUSSION THREAD

**Discussion by jjenko** [wantedly/frontend-night#260] [General]
## Policy

- 資料の準備をがんばらない
- 参加・途中退席は自由
- スキップしない

話したい・聞きたいネタを書いてく ✏️

ハッシュタグ [#frontend_night](https://twitter.com/hashtag/frontend_night?src=hash)

Prev: https://github.com/wantedly/frontend-night/discussions/259 

 _Created from https://github.com/wantedly/frontend-night/discussions/179 by [issue-creator](https://github.com/rerost/issue-creator)_

**Comment by chloe463** [comment]
## Nex.js の app router と React server components について

https://nextjs.org/docs/getting-started/react-essentials


- app router を使った場合、デフォルトで server components が使用される
- server components から import されるコンポーネントは app ディレクトリにいなくても server components として扱われる。
- これまでと同様の CSR されうるコンポーネントとして扱うには `"use client"` というディレクティブをファイル先頭に書く必要がある。
- React が server components を発表した当初には、server components と client components と区別には `foo.server.tsx` や `bar.client.tsx` といったファイル名にするとあったが、Next.js はファイル名での判断は行っていない様子。
- React server components では useState をはじめとする hooks や context は使えない。
- React server components から client components を使用することはできるが、client components から server components を使用することはできない。
   - children として server components を受け取って呼び出すことは可能


これは NG
```Foo.client.tsx
import { Bar } from "./bar.server.ts";

const Foo = () => {
   return (
      <div>
		<span>This component does not work well...</span>
         <Bar />
      </div>
   );
};
```

これは OK
```hoge.server.tsx
// Hoge.server.tsx
import { Bar } from "./bar.server.tsx";
const Hoge = () => {
   return (
       <Foo>
		<span>This works</span>
          <Bar />
       </Foo>
   );
}


// Foo.client.tsx
const Foo = ({ children }) => {
   return (
      <div>
         {children}
      </div>
   );
};
```

- App router (react server components) を使った開発では、基本的に server components を使用し、component tree でいうところの leaf (末端のコンポーネント) を client components にするべきという指針がNext.js公式ドキュメントには書かれている。
   - しかしながら、server components では styled-components が使用できないため、現状の Wantedly の技術スタックを維持しながらやろうとした場合かなり使用箇所が限られてしまうと考えている。
   - zero runtime な CSS in JS を使うとか、Tailwind を使うとか、styled-components が対応するのを(期待薄かも)待つか 🤔 
  - Server components で API リクエストすると、Apollo client cache との連動ができなくなるので、optimistic response と相性が悪いかもしれない
 - Server components は Node.js の関数が普通に使えるので、 environments あたりの仕組みをこれで吸収するという案はある。

**Comment by s-kawabe** [comment]
（小ネタ）
https://chrome.google.com/webstore/detail/locatorjs/npbfdllefekhdplbkdigpncggmojpefi

LocatorJS 使ってない人いたらおすすめです（local fork環境でも動く）

https://github.com/wantedly/frontend-night/assets/65433193/45bc66dc-005f-49d9-a903-e36eca1e7789
